src/diffusers/pipelines/consistency_models/pipeline_consistency_models.py

```diff
@@ -0,0 +1,395 @@
+        # TODO: how do we ensure that this in Karras sigma space rather than in "original" sigma space?
+        # 5. Get sigma schedule
+        assert hasattr(self.scheduler, "sigmas"), "Scheduler needs to operate in sigma space"
+        if hasattr(self.scheduler, "sigma_min"):
+            # Overwrite sigma_min with sigma_min from the scheduler
+            sigma_min = self.scheduler.sigma_min
+            sigma_max = self.scheduler.sigma_max
+        sigmas = self.scheduler.sigmas
+
+        # 6. Prepare extra step kwargs. TODO: Logic should ideally just be moved out of the pipeline
+        extra_step_kwargs = self.prepare_extra_step_kwargs(generator, eta, sigma_min, sigma_max)
+
+        # 7. Denoising loop
+        if num_inference_steps == 1:
+            # Onestep sampling: simply evaluate the consistency model at the first sigma
+            # See https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L643
+            sigma = sigma_max
+            sigma_in = sample.new_ones([sample.shape[0]]) * sigma
+            _, sample = self.denoise(
+                sample,
+                sigma_in,
+                class_labels=class_labels,
+                sigma_min=sigma_min,
+                sigma_data=sigma_data,
+                clip_denoised=clip_denoised,
+            )
+        else:
+            # Multistep sampling or Karras sampler
+            num_warmup_steps = len(timesteps) - num_inference_steps * self.scheduler.order
+            with self.progress_bar(total=num_inference_steps) as progress_bar:
+                for i, t in enumerate(timesteps):
+                    sigma = sigmas[i]
+                    sigma_in = sample.new_ones([sample.shape[0]]) * sigma
+
+                    # TODO: should we call scale_model_input here?
+                    sample = self.scheduler.scale_model_input(sample, t)
+                    model_output, denoised = self.denoise(
+                        sample,
+                        sigma_in,
+                        class_labels=class_labels,
+                        sigma_min=sigma_min,
+                        sigma_data=sigma_data,
+                        clip_denoised=clip_denoised,
+                    )
+
+                    # Works for both Karras-style schedulers (e.g. Euler, Heun) and the CM multistep scheduler
+                    sample = self.scheduler.step(denoised, t, sample, **extra_step_kwargs).prev_sample
+
+                    # Note: differs from callback support in original code
+                    # See e.g. https://github.com/openai/consistency_models/blob/main/cm/karras_diffusion.py#L459
+                    # call the callback, if provided
+                    if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                        progress_bar.update()
+                        if callback is not None and i % callback_steps == 0:
+                            callback(i, t, sample)
```

Comment:
Ok this is great! The main thing we need to do here is maintain the scheduler abstraction. I.e. some of the denoising is occurring in the pipeline in this case which is why we are munging with the sigmas in the pipeline which we shouldn't have to do. 

Maintaining the scheduler abstraction isn't necessarily because we expect to swap out other schedulers on the pipeline but more just for maintainability in this case.

I believe most of the manual munging that is being done can be handled by the existing scheduler api. I.e. in `denoise`, we are scaling the input to the unet but we already have a `scale_model_input` function on the scheduler. I'm not sure if we have an existing api for pipelines which are conditioned on sigmas instead of timesteps. I think it should be ok to let the pipeline directly index into `scheduler.sigmas` but do no more processing of the sigma in the pipeline (if this isn't possible and it we really need to do some additional processing of the sigma in the pipeline let me know pls!). 

We should also ideally be able to pass the unet output directly to the scheduler. This step `denoised = c_out * model_output + c_skip * x_t` can occur in the scheduler.

I think that should eliminate most of the sigma munging in the pipeline. Let me know if that tracks and I'll try to get a concrete answer to you on how to use the sigma in the pipeline :)

Reply:
https://github.com/huggingface/diffusers/blob/41ae6708280ba84a4671d20fbb4d52f176083faa/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_latent_upscale.py#L494 See here for an example of using sigmas in the pipeline, hopefully it shouldn't need to be more complicated than that :)

Reply:
Should have removed most of the sigma munging in the pipeline. See also https://github.com/huggingface/diffusers/pull/3492#issuecomment-1579975036.